go/vt/vtgate/vindexes/lookup_internal.go

```diff
@@ -73,21 +73,87 @@ func (lkp *lookupInternal) Lookup(vcursor VCursor, ids []sqltypes.Value) ([]*sql
 		return nil, fmt.Errorf("cannot perform lookup: no vcursor provided")
 	}
 	results := make([]*sqltypes.Result, 0, len(ids))
-	for _, id := range ids {
+	if !ids[0].IsIntegral() && !ids[0].IsBinary() {
+		// for non integral and binary type, fallback to send query per id
+		for _, id := range ids {
+			vars, err := sqltypes.BuildBindVariable([]interface{}{id})
+			if err != nil {
+				return nil, fmt.Errorf("lookup.Map: %v", err)
+			}
+			bindVars := map[string]*querypb.BindVariable{
+				lkp.FromColumns[0]: vars,
+			}
+			var result *sqltypes.Result
+			co := vtgatepb.CommitOrder_NORMAL
+			if lkp.Autocommit {
+				co = vtgatepb.CommitOrder_AUTOCOMMIT
+			}
+			result, err = vcursor.Execute("VindexLookup", lkp.sel, bindVars, false /* rollbackOnError */, co)
+			if err != nil {
+				return nil, fmt.Errorf("lookup.Map: %v", err)
+			}
+			rows := make([][]sqltypes.Value, 0, len(result.Rows))
+			for _, row := range result.Rows {
+				rows = append(rows, []sqltypes.Value{row[1]})
+			}
+			results = append(results, &sqltypes.Result{
+				Rows: rows,
+			})
+		}
+	} else {
+		// for integral or binary type, batch query all ids and then map them back to the input order
+		idVars := make([]interface{}, len(ids))
+		for i := range ids {
+			idVars[i] = ids[i]
+		}
+		vars, err := sqltypes.BuildBindVariable(idVars)
+		if err != nil {
+			return nil, fmt.Errorf("lookup.Map: %v", err)
+		}
 		bindVars := map[string]*querypb.BindVariable{
-			lkp.FromColumns[0]: sqltypes.ValueBindVariable(id),
+			lkp.FromColumns[0]: vars,
 		}
-		var err error
-		var result *sqltypes.Result
 		co := vtgatepb.CommitOrder_NORMAL
 		if lkp.Autocommit {
 			co = vtgatepb.CommitOrder_AUTOCOMMIT
 		}
-		result, err = vcursor.Execute("VindexLookup", lkp.sel, bindVars, false /* rollbackOnError */, co)
+		result, err := vcursor.Execute("VindexLookup", lkp.sel, bindVars, false /* rollbackOnError */, co)
 		if err != nil {
 			return nil, fmt.Errorf("lookup.Map: %v", err)
 		}
-		results = append(results, result)
+		resultMap := make(map[string][][]byte)
+		t := querypb.Type_NULL_TYPE
+		for _, row := range result.Rows {
+			t = row[1].Type()
+			val, ok := resultMap[string(row[0].Raw())]
+			if ok {
+				val = append(val, row[1].Raw())
+			} else {
+				val = [][]byte{row[1].Raw()}
+			}
+			resultMap[string(row[0].Raw())] = val
```

Comment:
I believe this section can be simplified to: `resultMap[string(row[0].ToString())] = append(resultMap[string(row[0].ToString())], row[1])`